pkg/ruler/ruler.go

```diff
@@ -993,7 +995,31 @@ func (r *Ruler) GetRules(ctx context.Context, req RulesRequest) ([]*GroupStateDe
 		return nil
 	})
 
-	return merged, err
+	// If the request asks for pagination, we fetch req.MaxGroups number
+	// of rule groups from each replica. We then merge and sort these and
+	// take the top k (k = MaxGroups)
+	if req.MaxGroups > 0 {
+		slices.SortFunc(merged, func(a, b *GroupStateDesc) int {
+			nsCmp := strings.Compare(a.Group.Namespace, b.Group.Namespace)
+			if nsCmp != 0 {
+				return nsCmp
+			}
+
+			// If Namespaces are equal, check the group names
+			return strings.Compare(a.Group.Name, b.Group.Name)
+		})
```

Comment:
Non-blocking nit: this might be over-engineering..... but I see a potential issue where someone updates the sort strategy in one place, and not the other (in api.go). Could we share them somehow? Define a "SortableGroup" interface or something?